File: app/src/main/java/com/n8yn8/abma/model/dao/EventDao.java
package com.n8yn8.abma.model.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.n8yn8.abma.model.entities.Event;
import com.n8yn8.abma.model.old.DatabaseHandler;

import java.util.List;

@Dao
public interface EventDao {

    @Query("SELECT * FROM events")
    List<Event> getEvents();

    @Query("SELECT * FROM events WHERE " + DatabaseHandler.KEY_YEAR_ID + "=:yearId ORDER BY start_date ASC")
    List<Event> getEvents(final String yearId);

    @Query("SELECT * FROM events WHERE " + DatabaseHandler.KEY_YEAR_ID + "=:yearId ORDER BY start_date ASC")
    LiveData<List<Event>> getEventsLive(final String yearId);

    @Query("SELECT * FROM events WHERE start_date >= :startDate AND start_date <= :endDate ORDER BY start_date ASC")
    LiveData<List<Event>> getAllEventsFor(long startDate, long endDate);

    @Query("SELECT * FROM events WHERE object_id = :objectId LIMIT 1")
    Event getEventById(String objectId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Event event);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(List<Event> event);

    @Query("SELECT * FROM events WHERE start_date < :startDate ORDER BY start_date DESC LIMIT 1")
    Event getEventBefore(long startDate);

    @Query("SELECT * FROM events WHERE year_id=:yearId AND start_date < :startDate ORDER BY start_date DESC LIMIT 1")
    Event getEventBefore(String yearId, long startDate);

    @Query("SELECT * FROM events WHERE start_date > :startDate ORDER BY start_date ASC LIMIT 1")
    Event getEventAfter(long startDate);

    @Query("SELECT * FROM events WHERE year_id=:yearId AND start_date > :startDate ORDER BY start_date ASC LIMIT 1")
    Event getEventAfter(String yearId, long startDate);

    @Delete
    void delete(Event... events);
}

File: app/src/main/java/com/n8yn8/abma/model/dao/EventDao.kt
package com.n8yn8.abma.model.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.n8yn8.abma.model.entities.Event
import com.n8yn8.abma.model.entities.EventPapers

@Dao
interface EventDao {
    @Query("SELECT * FROM events")
    suspend fun events(): List<Event>

    @Query("SELECT * FROM events WHERE year_id = :yearId ORDER BY start_date ASC")
    suspend fun getEvents(yearId: String): List<Event>

    @Query("SELECT * FROM events WHERE year_id = :yearId ORDER BY start_date ASC")
    fun getEventsLive(yearId: String): LiveData<List<Event>>

    @Query("SELECT * FROM events WHERE start_date >= :startDate AND start_date <= :endDate ORDER BY start_date ASC")
    fun getAllEventsFor(startDate: Long, endDate: Long): LiveData<List<Event>>

    @Transaction
    @Query("SELECT * FROM events WHERE object_id = :objectId LIMIT 1")
    suspend fun getEventById(objectId: String): EventPapers?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(event: Event)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(event: List<Event>)

    @Transaction
    @Query("SELECT * FROM events WHERE start_date < :startDate ORDER BY start_date DESC LIMIT 1")
    suspend fun getEventBefore(startDate: Long): EventPapers?

    @Query("SELECT * FROM events WHERE year_id = :yearId AND start_date < :startDate ORDER BY start_date DESC LIMIT 1")
    suspend fun getEventBefore(yearId: String, startDate: Long): Event?

    @Transaction
    @Query("SELECT * FROM events WHERE start_date > :startDate ORDER BY start_date ASC LIMIT 1")
    suspend fun getEventAfter(startDate: Long): EventPapers?

    @Query("SELECT * FROM events WHERE year_id = :yearId AND start_date > :startDate ORDER BY start_date ASC LIMIT 1")
    suspend fun getEventAfter(yearId: String, startDate: Long): Event?

    @Delete
    fun delete(vararg events: Event)
}